Maintain ipcache entries for Cilium host IPs based on k8s node annotations

aanm opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4797)
<!-- Reviewable:end -->

aanm committed:
add cilium-host IP to k8s node's annotation

This allow us to create a tunnel for host networking Pods when the
Cilium agent is not yet registered in the KVStore.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
k8s: Create ipcache entries for CiliumIP announce via node annotations

This allows routing and policy enforcement of the internal Cilium IP before the
kvstore is up if Kubernetes node annotations are available.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm committed:
daemon: GC ipcache for entries created by KVStore and/or agent-local

To prevent the premature deletion of the IPs created by kubernetes,
until the other agent is connected to the kvstore, Cilium should do
the GC for the IPs that were created exclusively by the kvstore
and/or agent-local.

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1632,3 +1675,56 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
+
+	n := k8s.ParseNode(k8sNode)
+	var hostIP net.IP
+	if n.Name != node.GetName() {
```
Any reason to continue for the own node? Leaving `hostIP` nil seems weird here

aanm replied:
fixed

tgraf replied:
Is this fixed? hostIP is still nil for the own node and nil is passed into `To4()`.

aanm replied:
@tgraf it is, but it was wrongly on the next commit. I've put the changes into this commit

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1632,3 +1675,56 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
+}
+
+func (d *Daemon) updateK8sNodeV1(_, k8sNode *v1.Node) {
+	// TODO rely on the kvstore integration?
```
This can just call the same code as add

aanm replied:
fixed

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1632,3 +1675,56 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
+}
+
+func (d *Daemon) deleteK8sNodeV1(k8sNode *v1.Node) {
+	// TODO rely on the kvstore integration?
```
This needs to do the same as for pod delete events:

```
        // a small race condition exists here as deletion could occur in
        // parallel based on another event but it doesn't matter as the
        // identity is going away
        id, exists := ipcache.IPIdentityCache.LookupByIP(pod.Status.PodIP)
        if !exists {
                return true, fmt.Errorf("identity for IP does not exist in case")
        }

        if id.Source != ipcache.FromKubernetes {
                return true, fmt.Errorf("ipcache entry not owned by kubernetes source")
        }

        ipcache.IPIdentityCache.Delete(pod.Status.PodIP)
```

aanm replied:
fixed

joestringer commented on daemon/k8s_watcher.go:
```diff
@@ -1632,3 +1675,120 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
+		logger.WithField("hostIP", hostIPStr).Debugf("ipcache entry not owned by kubernetes source")
+	}
+
+	ipcache.IPIdentityCache.Delete(hostIPStr)
```
This doesn't delete the entry from the underlying ipcache implementation, only from the `IPIdentityCache`. If this is the way that you want to manage the IPCache, then perhaps the `OnIPIdentityCacheGC()` above should not ignore the `FromKubernetes` case.

tgraf replied:
Let's just add
```
        idPair := identity.IPIdentityPair{IP: podIP}
        for _, listener := range d.ipcacheListeners {
                listener.OnIPIdentityCacheChange(ipcache.Delete, nil, idPair)
        }
```

for now and #4810 can refactor all of this

aanm replied:
Done

joestringer commented on daemon/k8s_watcher.go:
```diff
@@ -1632,3 +1675,120 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
+			}
+
+			// Delete previous cilium IP for the given node.
+			ipcache.IPIdentityCache.Delete(ciliumIPStrOld)
```
AFAICT `IPCache.Upsert()` should handle this case for you.

aanm replied:
I'm not calling `Upsert()` anywhere. `OnIPIdentityCacheChange` doesn't call `Upsert` as well. Line 1589 is doing the same. cc @tgraf why aren't we calling `IPCache Update`?

tgraf replied:
@joestringer is right. The call to `Upsert()` on line 1744 will handle both insertions and updates. It will delete and re-add the entry if needed and also skip if the mapping is already identical. I think you can remove all code to detect updates.

joestringer commented on daemon/k8s_watcher.go:
```diff
@@ -1632,3 +1675,120 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
+	}
+
+	updated := ipcache.IPIdentityCache.Upsert(ciliumIPStrNew, ipcache.Identity{
+		ID:     identity.ReservedIdentityCluster,
```
Oh, interesting. Up until now, cluster has meant exactly one thing: On egress, the destination lies within the "cluster prefix" (typically 10.0.0.0/8) and is not otherwise managed by Cilium.

This now gives new meaning to cluster, to say it can also mean other nodes where Cilium is operating. Previous proposals [#3536, #4779] have proposed a new identity for this. I think we would benefit from actually having a new, different identity for this rather than making the meaning of "cluster" more complicated (unless you have a strong argument for what "cluster" should mean in future releases).

Note that this also introduces a potential policy incompatibility: IPs with the same role in Cilium 1.1 will now be classified as a different identity in Cilium 1.2. Therefore we should consider a `host-allows-world` style option for backwards compatibility. [[host-allows-world](https://cilium.link/host-vs-world)]

joestringer replied:
For reference, currently, `reserved:cluster` matches the agent CLI `--ipv4-node` IP using the agent CLI `--ipv4-cluster-cidr-mask-size` CIDR range, ie by default it means `10.0.0.0/8` traffic and nothing else.

tgraf replied:
We can map it to ReservedHost for now as it is via the kvstore and then discuss introducing a new identity in a separate PR.

aanm replied:
Changed to `ReservedIdentityHost`

joestringer reviewed (commented):
No IPv6?

I don't know if this is viable, but I think that this PR would very much benefit from the refactoring in #4810.

I've raised some questions below in the meaning of the cluster identity which I think needs some serious consideration.

joestringer commented on daemon/k8s_watcher.go:
```diff
@@ -1632,3 +1675,120 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
+			if identityPairOld.IP.Equal(identityPairNew.IP) &&
+				identityPairOld.HostIP.Equal(identityPairNew.HostIP) {
+				// If both identities are the same, there's no point on
+				// continuing doing any operation.
```
The above checks the IPs only. What if the Identity is different? Could that be the case, perhaps even if only during upgrade?

aanm replied:
@joestringer I don't understand your concern. The identity is something that is not transfered between agents. Between upgrades the cilium-agent will only receive "adds" in the beginning, only latter on it will receive "updates"

joestringer replied:
OK, that answers my question.

tgraf commented on daemon/k8s_watcher.go:
```diff
@@ -1632,3 +1675,134 @@ func (d *Daemon) deleteK8sPodV1(pod *v1.Pod) {
+	identityPairNew, ciliumIPStrNew, err := getIDs(nodeNew, k8sNodeNew)
+	if err != nil {
+		return err
+	}
```
There are several possible nil dereference of `identityPairNew` below as `getIDs()` can return nil:
```
+	getIDs := func(node *node.Node, k8sNode *v1.Node) (*identity.IPIdentityPair, string, error) {
+		if node == nil {
+			return nil, "", nil
+		}
+		hostIP := node.GetNodeIP(false)
+		if ip4 := hostIP.To4(); ip4 == nil {
+			logger.Debugf("Skipping ipcache update, HostIP is not an IPv4 address %s", hostIP)
+			return nil, "", nil
+		}
```

I think you want to bail out if `getIDs()` returns nil

aanm replied:
Done

tgraf committed:
k8s: Update HostIP even if identity is unchanged

The current code made ipcache.Upsert() return false if the update was skipped.
This resulted in OnIPIdentityChange() to not be invokved if the identity was
unchanged. Instead, have Upsert() indicate whether the entry is owned by
another owner to properly skip updates.

Fixes: 82a7ce9bad0 ("ipcache: Populate ipcache based on Pod events")

Signed-off-by: Thomas Graf <thomas@cilium.io>
